File: Strings/removeKdigits.cpp
#include <bits/stdc++.h>
using namespace std;

string removeKdigits(string num, int k)
{

    int n = num.length();
    if (n == k)
        return "0";
    if(k>n)
        return "0";
    
    stack<char> st;
    int i = 0;

    for(int i=0;i<n;i++)
    {
        while (k > 0 && !st.empty() && st.top() > num[i])
        {
            st.pop();
            k--;
        }
        st.push(num[i]);
    }

    while (k > 0)
    {
        st.pop();
        k--;
    }

    string ans = "";

    while (!st.empty())
    {
        ans += st.top();
        st.pop();
    }

    reverse(ans.begin(), ans.end());

    while (ans.length() > 1 && ans[0] == '0')
    {
        ans.erase(ans.begin());
    }

    return ans;


}
int main()
{

    string str="112";
    int k=1;
    cout<<removeKdigits(str,k)<<endl;

    return 0;
}

File: LeetCode/Arrays/removeKDigits.java
import java.util.Stack;

class Solution {
    public String removeKdigits(String num, int k) {

        if (num.length() == k)
            return "0";
        
        Stack<Character> stack = new Stack<>();
        int i = 0;
        
        while (i < num.length()) {
            while (k > 0 && !stack.isEmpty() && stack.peek() > num.charAt(i)) {
                stack.pop();
                k--;
            }
            stack.push(num.charAt(i));
            i++;
        }

        // if k is not zero, then pop the remaining elements from the stack
    //    eg = 12345, k = 2
        while (k > 0 && !stack.isEmpty()) {
            stack.pop();
            k--;
        }
        
        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) {
            sb.append(stack.pop());
        }
        sb.reverse();
        
        // remove the leading zeros
        while (sb.length() > 1 && sb.charAt(0) == '0') {
            sb.deleteCharAt(0);
        }
        return sb.toString();
    }
}
public class removeKDigits {
    public static void main(String[] args) {
        String num = "001";
        int k = 3;
        Solution obj = new Solution();
        String ans = obj.removeKdigits(num, k);
        System.out.println(ans);
    }
}
